### DISCUSSION TITLE
Lower bound generic constraint

### DISCUSSION THREAD

**Discussion by fa1ntl1ght** [dotnet/csharplang#5377] [Language Ideas]
### Summary
Allow
```cs
void Method(ICollection<T> coll, int element) where int : T
{
	coll.Add(element);
}
```
### Explanation
We already have contravariance support for interfaces, so `IComparer<in T>` etc. is possible. However, generic constraints still lack proper contravariance support.
If we had to define a method that adds int values to a list we would expect it to support (suppose we have preview features enabled) `List<INumber>`, `List<IBinaryNumber>` and other collections whose type argument is an interface that is implemented by Int32.
It is currently possible to express contravariance through a generic parameter:
```cs
void Method<TColl, TLowerBound>(ICollection<TColl> coll, TLowerBound element) where TLowerBound : TColl
{
	coll.Add(element);
}
```
and call that method like so:
```cs
List<object> list = new();
Method(list, 5);
```
and this will work fine until we want to make `TLowerBound` a specific type.
So this isn't useful if you want to constrain a parameter to some actual type instead of an another parameter.
So, for symmetry with
`T : ActualType`
and
`T1 : T2`
and interfaces (`in` & `out`)
it would be logical to also add
`ActualType : T`
so all cases are covered.

Relevant links:
https://docs.oracle.com/javase/tutorial/java/generics/lowerBounded.html
https://github.com/microsoft/TypeScript/issues/14520

**Comment by HaloFour** [comment]
Using generic type constraints to represent contravariance is interesting.  While the relationship is currently covariant I don't think that was necessarily the intent, it just happens to be the only direction in which variance can be used to enable functionality of the generic type argument.  A contravariant relationship doesn't enable you to make any assumptions regarding the capabilities of the type, you can't call any members on it, etc.

I have to question as to whether trying to change that relationship more than 15 years in is worthwhile.  These problems are simply solved differently in C# than they are in Java or TypeScript, which have more flexible generics because they are erased at runtime and everything is effectively `Object` (or the lower bound).

I will note that your example cannot work as value types cannot be variant.  In Java primitive types simply can't participate with generics, but that will be addressed with Project Valhalla but with the same limitations of C# and the loss of variance for primitive/inline types.

**Reply by fa1ntl1ght** [reply]
If by example you mean this:
```cs
void Method(ICollection<T> coll, int element) where int : T
{
	coll.Add(element);
}
```
Why this shouldn't work? Does it have something to do with the runtime limitations? Since value types can implement interfaces and this kind of relationship between types can already be expressed with generic parameters, I don't see why this couldn't work with an actual type and a generic parameter. Value types are even treated as derived from ValueType and object at runtime. Am I missing something?

**Reply by HaloFour** [reply]
> Why this shouldn't work?

Value types require different storage and cannot be used directly as the interface without boxing.  There are some runtime features which can avoid the allocation but that wouldn't work across variance.

In Java you simply can't use a primitive/value type as a generic type argument, so all generic type arguments are reference types which enables the described variance.  If Valhalla does deliver specialization of generics with value types that will no longer be the case and there will be similar restrictions enforced by the runtime (as generics could no longer be fully erased).

> Value types are even treated as derived from ValueType and object at runtime.

That relationship isn't real, it's faked by the runtime to make value types feel like they share in a single type hierarchy.

**Reply by fa1ntl1ght** [reply]
> Value types require different storage and cannot be used directly as the interface without boxing. There are some runtime features which can avoid the allocation but that wouldn't work across variance.

> That relationship isn't real, it's faked by the runtime

Yes, I know all this. I am more interested in how this relationship works in this example:
```cs
void Method<TColl, TLowerBound>(ICollection<TColl> coll, TLowerBound element) where TLowerBound : TColl
{
	coll.Add(element);
}
List<object> list = new();
Method(list, 5);
```
Since we are basically calling `Method<object, int>` and `TLowerBound : TColl` becomes `int : object` it should be the same, isn't it?
Though I just looked into how these constraints are represented in dnspy (there is only a single direction of variance) and I see that runtime changes are required to make this work and I now partially understand what you talked about in your first comment. But I still don't understand what you told about value types. Can you elaborate more on "value types cannot be variant"? Since value types can't inherit from anything, there should not be any variance except for interfaces they implement which seems to be already supported through boxing/unboxing. What is the problem with making it work through boxing like covariance currently works?
